Mixing CRTP

brunoguindani opened the pull request:
I was fiddling with the CRTP for mixings, and I noticed that, if we implement it, then we have to pass template arguments whenever we perform base-to-derived conversion through `std::dynamic_pointer_cast()`, such as `mixing --> cond_mixing` in `ConditionalAlgorithm`.
We may handle this directly and somehow pass arguments, although no way for doing it comes to mind. If this is indeed impossible, we would have to move the intermediate `ConditionalMixing` and `MarginalMixing` up the inheritance tree (down the inheritance tree?), between `AbstractMixing` and `BaseMixing`, unlike what we did for hierarchies, so that both of them are not template classes and pointer conversion may resolve. Therefore we would need two "base" classes deriving from each of them: `ConditionalBaseMixing` and `MarginalBaseMixing`. It's quite the clunky design choice, so I wanted to discuss it with you.

Is there any better way?

P.S. Also, do we actually need `Mixing::clone()` since there's only one mixing per algorithm anyway?

mberaha commented:
I don't have a clear Idea on top of my head...

The 'easiest' solution would be to just move the virtual methods from Marginal and Conditional Mixing to `AbstractMixing` without having them purely virtual, but raise an error if one of these methods is called by default. Then the Mixing classes implementations only need to implement the ones they need.

I'm kind of pessimistic on the possibility of casting stuff around without keeping track of the template types, so I suggest we look for other solutions.

Also the idea of moving Conditional and Marginal mixing classes is kind of scary, since it would lead to two different BaseMixing classes depending on what you inherit from, so 2 more CRTPs to maintain (which is not the easiest pattern to debug to be honest).

brunoguindani commented:
I too dislike the split into two base classes.

> The 'easiest' solution would be to just move the virtual methods from Marginal and Conditional Mixing to `AbstractMixing` without having them purely virtual, but raise an error if one of these methods is called by default. Then the Mixing classes implementations only need to implement the ones they need.

Would your suggestion get rid of the intermediate classes? Is there any way to salvage them? Leaving compliance with a (pseudo-)"inheritance" structure to the user is not amazing, NGL.

mberaha commented:
TBH I don't know how we can have both the CRTP, the intermediate classes and be able to cast stuff around... Probably it's worth asking on Stackoverflow but I'm quite sure that it's going to be impossible

brunoguindani commented:
Here we go. I also corrected a small bug with a progress bar, which used number of iterations instead of chain size.

Lastly, I went and applied a temporary fix to grid methods, which still accept a grid of hierarchy covariates, even though we updated the rest of the evaluation methods to work with a single covariate. I'd rather fix them in a separate PR, since this one compiles and runs correctly anyway, at least as far as the three tests in `run_test.sh` go.

mberaha reviewed (approved):
Looks good to me! About the clone method in `BaseMixing`, I think we can safely assume that only one mixing is present at any point, so no cloning is needed.
